File: frontend/apps/gameboard-app/src/app/game-invites/game-invite.service.ts
// HTTP-backed client for the game-invites backend (gameboard/backend/gameboard
// game_invite.go + game_invite_service.go, served at /v0/api4gameboard/
// game-invites/*, already mounted through the wildcard route sneat-go's
// gameboard module registers — see backstage/docs/roadmaps/
// gameboard-game-invites.md §8.2 Phase 1).
//
// This REPLACES localStorage as this feature's persistence: the same-shaped
// prototype functions in game-invite-store.ts (createGameInvite,
// getGameInvite, listGameInvites, addRosterPlayer, setRsvp) are kept
// UNCHANGED in that file (never delete code) but are no longer imported by
// the organize/roster/rsvp/my-games pages — this service is. Mirrors
// GameService (../game.service.ts) for the sibling two-team scoreboard game:
// SneatApiService carries the signed-in user's Firebase ID token for writes
// that require one; public reads use getAsAnonymous.
import { inject, Injectable } from '@angular/core';
import { SneatApiService } from '@sneat/api';
import { firstValueFrom } from 'rxjs';
import { GameInviteDoc, RsvpStatus } from './game-invite-contract';
import { CreateGameInviteInput } from './game-invite-store';

/**
 * True when an error is SneatApiService's synchronous "no Firebase ID token"
 * guard, thrown before any network call. Same substring match GameService
 * uses (game.service.ts isNotAuthenticatedError) for the same reason: it's
 * the literal string `@sneat/api`'s errorIfNotAuthenticated() throws.
 */
function isNotAuthenticatedError(err: unknown): boolean {
  return typeof err === 'string' && err.includes('not authenticated');
}

/** Body of POST .../game-invites and .../game-invites/{id}/roster — a roster
 * player before a playerId is assigned. */
interface RosterPlayerBody {
  readonly name: string;
  readonly jersey?: string;
  readonly guardianName?: string;
}

/** Body of POST .../game-invites/by-token/{token}/rsvp. */
interface SubmitRsvpBody {
  readonly playerId: string;
  readonly status: RsvpStatus;
  readonly respondedBy: string;
  readonly note?: string;
}

/** Response of GET .../game-invites/by-token/{token} — the game (full public
 * read) plus which roster player, if any, the link specifically targets. */
export interface GameInviteByTokenResponse {
  readonly game: GameInviteDoc;
  readonly targetPlayerId?: string;
}

@Injectable({ providedIn: 'root' })
export class GameInviteService {
  private readonly api = inject(SneatApiService);

  /**
   * Organize a new game/practice for the caller's own team.
   * `POST /v0/api4gameboard/game-invites` — AUTHENTICATED (the backend stamps
   * organizerUID from the bearer token; an anonymous caller gets 401, so the
   * organize-game page must gate this call on sign-in — see
   * organize-game-page.component.ts's isSignedIn()/signIn(), mirroring
   * new-game-page's anon-first-then-gated-create pattern).
   */
  public createGameInvite(input: CreateGameInviteInput): Promise<GameInviteDoc> {
    return firstValueFrom(
      this.api.post<GameInviteDoc>('api4gameboard/game-invites', {
        sport: input.sport,
        teamName: input.teamName,
        opponentName: input.opponentName,
        scheduledMs: input.scheduledMs,
        venue: input.venue,
        playersNeeded: input.playersNeeded,
        recurring: input.recurring,
        organizerName: input.organizerName,
        roster: input.roster,
      }),
    );
  }

  /**
   * Read a game invite by id (the organizer/roster console read).
   * `GET .../game-invites/{id}` — PUBLIC/no-login (account-gate: "reads are
   * free"), so it uses getAsAnonymous.
   */
  public getGameInvite(gameId: string): Promise<GameInviteDoc> {
    return firstValueFrom(
      this.api.getAsAnonymous<GameInviteDoc>(
        `api4gameboard/game-invites/${encodeURIComponent(gameId)}`,
      ),
    );
  }

  /**
   * Add one player to an existing game invite's roster.
   * `POST .../game-invites/{id}/roster` — PUBLIC/anon-friendly by design: both
   * the roster console's "add a player" and the RSVP page's "not on the
   * list? add my kid" call this with no sign-in (mirrors the backend's
   * addRosterPlayer handler doc comment).
   */
  public addRosterPlayer(
    gameId: string,
    player: RosterPlayerBody,
  ): Promise<GameInviteDoc> {
    return firstValueFrom(
      this.api.postAsAnonymous<GameInviteDoc>(
        `api4gameboard/game-invites/${encodeURIComponent(gameId)}/roster`,
        player,
      ),
    );
  }

  /**
   * Resolve an invite-link token to the game it points at (+ which roster
   * player, if any, it targets). `GET .../game-invites/by-token/{token}` —
   * PUBLIC/anonymous, the RSVP page's entry read.
   */
  public getGameInviteByToken(
    token: string,
  ): Promise<GameInviteByTokenResponse> {
    return firstValueFrom(
      this.api.getAsAnonymous<GameInviteByTokenResponse>(
        `api4gameboard/game-invites/by-token/${encodeURIComponent(token)}`,
      ),
    );
  }

  /**
   * Submit a parent-proxy RSVP against a link token: `respondedBy` (the
   * PARENT, the actor) is always distinct from `playerId` (the CONTACT, the
   * kid the response is about) — see parent-proxy.ts.
   *
   * `POST .../game-invites/by-token/{token}/rsvp` — ANONYMOUS-FRIENDLY: works
   * with no session (the whole point of the anon-first RSVP page). If the
   * responder IS signed in, the request is sent authenticated first so the
   * backend can enrich the participant index (their "my games" list picks
   * this game up); an absent/expired session falls back to an anonymous post
   * — same try/fallback shape as GameService.appendEvent.
   */
  public async submitRsvpByToken(
    token: string,
    body: SubmitRsvpBody,
  ): Promise<GameInviteDoc> {
    const path = `api4gameboard/game-invites/by-token/${encodeURIComponent(token)}/rsvp`;
    try {
      return await firstValueFrom(this.api.post<GameInviteDoc>(path, body));
    } catch (err) {
      if (!isNotAuthenticatedError(err)) {
        throw err;
      }
      return firstValueFrom(
        this.api.postAsAnonymous<GameInviteDoc>(path, body),
      );
    }
  }

  /**
   * List the games the caller organizes + has RSVP'd to while signed in.
   * `GET /v0/api4gameboard/game-invites` — AUTHENTICATED (an anonymous
   * visitor has no stable identity to list "my games" against); callers
   * should gate this on isSignedIn() rather than let it throw
   * "not authenticated" (see game-invites-list-page.component.ts).
   */
  public listMyGameInvites(): Promise<GameInviteDoc[]> {
    return firstValueFrom(
      this.api.get<GameInviteDoc[]>('api4gameboard/game-invites'),
    );
  }
}

File: backend/gameboard/game_invite_service.go
package gameboard

import (
	"context"
	"errors"
	"strings"
)

// ErrGameInviteStoreUnconfigured is returned when the wired store does not
// implement GameInviteStore (mirrors "game store not configured" in service.go).
var ErrGameInviteStoreUnconfigured = errors.New("gameboard: game-invite store not configured")

// GameInviteByTokenResult is the resolved read for an anonymous invite-link
// visit: the game (teaser + full roster/RSVP state, all public reads per
// account-gate's "reads are free") plus which roster player, if any, the link
// specifically targets (role-invites' targeted-invite pattern; empty for an
// open join link).
type GameInviteByTokenResult struct {
	Game           GameInviteRecord
	TargetPlayerID string
}

// CreateGameInvite creates a new game-invite aggregate — the "organize a
// game" step (roadmap doc MVP item 1). organizerUID is the authenticated
// caller, stamped server-side (never client-supplied), mirroring
// Service.CreateGame's createdBy.
func (s *Service) CreateGameInvite(ctx context.Context, organizerUID string, input CreateGameInviteInput) (GameInviteRecord, error) {
	if s.gameInvites == nil {
		return GameInviteRecord{}, ErrGameInviteStoreUnconfigured
	}
	teamName := strings.TrimSpace(input.TeamName)
	if teamName == "" {
		return GameInviteRecord{}, ErrInvalidGameInvite
	}
	organizerName := strings.TrimSpace(input.OrganizerName)
	if organizerName == "" {
		organizerName = "Coach"
	}
	roster := make([]RosterPlayer, 0, len(input.Roster))
	for _, p := range input.Roster {
		name := strings.TrimSpace(p.Name)
		if name == "" {
			continue // a blank draft row is silently dropped, mirroring the
			// frontend's addPlayer guard (draftName().trim() must be non-empty
			// before the button is even enabled).
		}
		roster = append(roster, RosterPlayer{
			PlayerID:     newGameID(), // generic dashless-hex id generator (game.go), reused for roster player ids too
			Name:         name,
			Jersey:       strings.TrimSpace(p.Jersey),
			GuardianName: strings.TrimSpace(p.GuardianName),
		})
	}
	now := nowFunc().UnixMilli()
	g := GameInviteRecord{
		GameInviteID:  newGameID(),
		Sport:         input.Sport,
		TeamName:      teamName,
		OpponentName:  strings.TrimSpace(input.OpponentName),
		ScheduledMs:   input.ScheduledMs,
		Venue:         strings.TrimSpace(input.Venue),
		PlayersNeeded: input.PlayersNeeded,
		Recurring:     input.Recurring,
		OrganizerName: organizerName,
		OrganizerUID:  organizerUID,
		Roster:        roster,
		Responses:     map[string]RsvpRecord{},
		CreatedAtMs:   now,
		UpdatedAtMs:   now,
	}
	if err := s.gameInvites.CreateGameInvite(ctx, g); err != nil {
		return GameInviteRecord{}, err
	}
	return g, nil
}

// GameInvite reads a game-invite record by id (the organizer/roster console
// read) — public, no auth required (account-gate: "reads are free").
func (s *Service) GameInvite(ctx context.Context, gameInviteID string) (GameInviteRecord, error) {
	if s.gameInvites == nil {
		return GameInviteRecord{}, ErrGameInviteNotFound
	}
	return s.gameInvites.GetGameInvite(ctx, gameInviteID)
}

// AddRosterPlayer adds one player to an existing game invite's roster.
// PUBLIC/anon-friendly by design (see Handler.addRosterPlayer's doc comment):
// both the organizer console's "add a player" and the RSVP page's "not on the
// list? add my kid" call this today with no sign-in required.
func (s *Service) AddRosterPlayer(ctx context.Context, gameInviteID string, player CreateRosterPlayerInput) (GameInviteRecord, error) {
	if s.gameInvites == nil {
		return GameInviteRecord{}, ErrGameInviteStoreUnconfigured
	}
	name := strings.TrimSpace(player.Name)
	if name == "" {
		return GameInviteRecord{}, ErrInvalidGameInvite
	}
	rp := RosterPlayer{
		PlayerID:     newGameID(),
		Name:         name,
		Jersey:       strings.TrimSpace(player.Jersey),
		GuardianName: strings.TrimSpace(player.GuardianName),
	}
	return s.gameInvites.AddRosterPlayer(ctx, gameInviteID, rp)
}

// GameInviteByToken resolves an anonymous invite-link visit: decodes the
// token (never requires auth — anyone with the link may read the teaser/
// roster/fill), reads the game, and reports which roster player the link
// targets (if any). Mirrors eventius's links.Resolve(token) → event read
// (rsvp_handlers.go submitRsvp) but combined into one call since the
// game-invites frontend always needs both together.
func (s *Service) GameInviteByToken(ctx context.Context, token string) (GameInviteByTokenResult, error) {
	if s.gameInvites == nil {
		return GameInviteByTokenResult{}, ErrGameInviteNotFound
	}
	gameID, playerID, ok := decodeInviteToken(token)
	if !ok {
		return GameInviteByTokenResult{}, ErrInvalidToken
	}
	g, err := s.gameInvites.GetGameInvite(ctx, gameID)
	if err != nil {
		return GameInviteByTokenResult{}, err
	}
	return GameInviteByTokenResult{Game: g, TargetPlayerID: playerID}, nil
}

// SubmitRsvpByToken records a parent-proxy RSVP against a link token.
// ANONYMOUS-FRIENDLY: respondedByUID is "" for an anonymous submission and
// only set when the HTTP caller supplied a valid bearer token (see
// Handler.submitRsvpByToken) — mirrors eventius's ViaAccount (rsvp.go).
// Idempotent per (gameID, playerID): a resubmission for the same player
// overwrites rather than duplicates (SubmitRsvp's map-key semantics).
func (s *Service) SubmitRsvpByToken(ctx context.Context, token, respondedByUID string, input SubmitRsvpInput) (GameInviteRecord, error) {
	if s.gameInvites == nil {
		return GameInviteRecord{}, ErrGameInviteStoreUnconfigured
	}
	gameID, _, ok := decodeInviteToken(token)
	if !ok {
		return GameInviteRecord{}, ErrInvalidToken
	}
	switch input.Status {
	case RsvpGoing, RsvpMaybe, RsvpOut:
	default:
		return GameInviteRecord{}, ErrInvalidGameInvite
	}
	playerID := strings.TrimSpace(input.PlayerID)
	if playerID == "" {
		return GameInviteRecord{}, ErrInvalidGameInvite
	}
	respondedBy := strings.TrimSpace(input.RespondedBy)
	if respondedBy == "" {
		respondedBy = "A parent" // never attribute a response to an empty string (mirrors parent-proxy.ts buildProxyResponse)
	}
	rsvp := RsvpRecord{
		PlayerID:       playerID,
		Status:         input.Status,
		RespondedBy:    respondedBy,
		RespondedByUID: respondedByUID,
		RespondedAtMs:  nowFunc().UnixMilli(),
		Note:           strings.TrimSpace(input.Note),
	}
	updated, err := s.gameInvites.SubmitRsvp(ctx, gameID, rsvp)
	if err != nil {
		return GameInviteRecord{}, err
	}
	// Best-effort participant-index write for the "invited" half of "my
	// games" — never fails the RSVP itself (already durably persisted above),
	// mirrors eventius's attendance-edge enrichment being best-effort
	// (rsvp_handlers.go submitRsvp: "enrichment is best-effort").
	if respondedByUID != "" {
		_ = s.gameInvites.PutGameInviteParticipant(ctx, respondedByUID, gameID)
	}
	return updated, nil
}

// MyGameInvites lists the games uid organizes plus the games uid has RSVP'd
// to while signed in (the participant index), de-duplicated. AUTHENTICATED
// only (Handler.listMyGameInvites requires a bearer token) — an anonymous
// visitor has no stable identity to list "my games" against.
func (s *Service) MyGameInvites(ctx context.Context, uid string) ([]GameInviteRecord, error) {
	if s.gameInvites == nil {
		return nil, nil
	}
	organized, err := s.gameInvites.ListGameInvitesByOrganizer(ctx, uid)
	if err != nil {
		return nil, err
	}
	invited, err := s.gameInvites.ListGameInvitesByParticipant(ctx, uid)
	if err != nil {
		return nil, err
	}
	seen := make(map[string]bool, len(organized))
	out := make([]GameInviteRecord, 0, len(organized)+len(invited))
	for _, g := range organized {
		seen[g.GameInviteID] = true
		out = append(out, g)
	}
	for _, g := range invited {
		if seen[g.GameInviteID] {
			continue
		}
		seen[g.GameInviteID] = true
		out = append(out, g)
	}
	return out, nil
}
